### ISSUE TITLE
some CH preparations problems 

### ISSUE THREAD

**Issue by FallenKain** [graphhopper/graphhopper#832]
i watch changes in master version of PrepareCH class and it's almost same that's used in 0.5 version(that I use), that's why i try to explain what problems i find in this case :)
1) shortcuts logic has a big problem that algo did only two variants of changing - create or modify already created, and that's wrong, that's why:
we have road, that has quite little detour

--1 ----2---3--
    \          /
     4-----5

way 1-2-3 more effective, that way 1-4-5-3, but what will happens if way 1-2-3, for example, blocked from 05:00AM to 08:00AM? For point's 1 and 3 created shortcut what has links to ways 1-2 and 1-3 coz less weighting and when we try to pass this part of road we have 2 variants: a) we can't manually block 1-2-3 ways in routing algo, then we has output that contains 1-2 and 2-3 ways - error b) we can manually block this edges, but ch algo try to go by shortcuts and we have detour out of this part of road graph - again error

If CH preparer could create shortcut's for same start and end nodes, but different edges(not for all but with some weighting difference), then we may go to path 1-4-5-3

2) second problem with ferry way's nodes and roads near it on docks. Ferry ways nodes and nodes on docks roads has low priority(in most cases it has only 2 edges from node) and some times CH preparer set wrong nodes levels, coz level depends on node id(that's wrong logic, isn't?) when creates shortcuts, in result we can't ride to ferry ways by CH algo(even if this is only one way from start to end), for example way from point=49.05407,140.151215 to point=46.946512,142.68631 must go through https://www.openstreetmap.org/way/331374594 this ferry way, but levels not right and algo can't do this, so detour far away.

**Comment by karussell** [comment]
I'm currently a bit lost with your explanation :) ... do you have a unit test or integration test which shows the problem?

> blocked from 05:00AM to 08:00AM?

We do not support time dependency, or what do you mean here?

> set wrong nodes levels, coz level depends on node id

Why do you think so?

> must go through this ferry way, but levels not right and algo can't do this, so detour far away.

Why do you think it is a CH issue? When you disable CH and do no preparation is the ferry taken? Isn't it more an access problem somewhere on the ferry? CH can of course use ferry if allowed for the vehicle.

**Comment by FallenKain** [comment]
first it's something like "you can't ride best way but way still in graph", and no test's for this, i'll try to do some modifications in this class like creating new shortcuts when we have same point's but quite a little difference between weights, but this changed levels of nodes and algo sometimes start detour edges, when it must go trough :)
second - i debug this, and where was not expected nodes level, just because one node id was less that another, they has same priority, and creations shortcuts on one node effected another node, if revert id's order - this part of way was ok :) 
And I have question, why do you do so realization, where nodes level can't be same? O_o

**Comment by karussell** [comment]
Sorry, I'm again unable to fully understand what you say, of course trying my bests ... just in case I misunderstood something :)

> first it's something like "you can't ride best way but way still in graph"

What do you mean here?

> depends on priority, but when priority same - set first that cames first, isn't it?

This is not problematic as we can use ANY order. In fact we build the levels via a heuristic, no need for an optimal solution as long as we create the necessary shortcuts.

> i debug this, and there was not expected nodes level

I really need an example for this. Also you can verify if you this is correct via enabling vs. disabling CH.

> if i right understand this class used by CH algo's EdgeFilter, and this class can't accept such edge, coz one node level highter that we expected :)

Maybe here is a misunderstanding. We are 'just' following the papers for contraction hierarchy, the normal shortest path algorithm do not need all this 'CH' stuff.

**Comment by FallenKain** [comment]
> first it's something like "you can't ride best way but way still in graph"

oh, i know, for example ways with tags access:destination or ways with emergency tag - we can't drop this ways from graph, coz in some cases it can be used for routing, but in most cases this ways must be droped from routing (don't have better example, have to much modifications in project that changes to much logic of your's original code)

example of second can't give coz differents tags are used in your's original project and in mine :) not all nodes that i have in my graph you have in your, that's why contraction difference and routings has diffs too, you can just write it by comment to somewhere that this may happen when priorities are the same :)

**Comment by karussell** [comment]
> we can't drop this ways from graph, coz in some cases it can be used for routing

Yes, sure. But the CH preparation is not designed for per-request changes, so this is not a bug.

> example of second can't give coz differents tags are used in your's original project and in mine

But do you still think there is a bug in the preparation? If there is a bug then it should be there regardless of the tagging and it should be reproduceable via unit or integration test

**Comment by FallenKain** [comment]
hm, i don't know how to classify correctly, coz this preparation can emit wrong routing logic, but on another hand it's not preparation bug, let's call this suggestion, i think, suggestion to add into original shortcust's (that created now) info about small (by weighting) detour if it exist, than, for example, if default routing path not correct for us, we can check shortcuts info's and find this detour or something like this :)

yes it was preparation bug, but there is to much variables that effected one by one, and so rarely(i have this problem only on couple nodes), that create test only for this variant was not good idea, coz i must to save my dump for this test (dump was 2.8 Gb)

**Comment by karussell** [comment]
Just turn off CH and you can do what you suggested. Otherwise it could lead to a buggy system if not done properly.

> yes it was preparation bug

Please try to explain it better with some examples, I do not think it is the case from what I've understood so far. Also test always with and without CH to make sure the problem is really in CH.

**Comment by FallenKain** [comment]
CH's shortcuts is too good for most cases, that's why turning this off highly undesirable :)

I use two algos(and yes, if don't use shortcuts all works fine) and debug this problem, for test you must have my car flag encoder, that has own dependences and they has own and so on... so i must give you my project code, that i can't do(sorry :( ), my dump (only dump may not emit this bug, coz nodes count, may be not same, contraction may not did same preparation) and config values for CH preparer, that i use... all of this give us only one - if someone find not routable edge by CH algorithm, it may be this bug

**Comment by karussell** [comment]
I still do not think this is a bug and there is a misunderstanding, you are looking to make CH more flexible, which is possible but not a simple task and requires to digg into the latest research. Have renamed the title accordingly. Also please see the issues #780 and #493 

If you still think there is a bug in CH (and not e.g. in your flagencoder etc) we need a reproduceable unit or integration test or a very good guess where we are doing something wrong directly in the code.

**Comment by FallenKain** [comment]
not CH preparation, all system of CH algo, I think that's right, and i do some working variant by editing edge filter, but still it's not so good, coz performance greatly degradate on it (coz some modifications was done with block filters at all), that's why if we have all needed information in shortcut or maybe create second shorcuts holder file, where id of detour shortcut will be same of id of original shortcut, then we can choose alternative variant of shortcut in edge filter if need :)

yeap i still think there, but this is hard to recreate by unit test, to many variables, that effected each other :/ but maybe one day when have not so much work i will try to do it 4 you :)
